tensilelite/Tensile/Components/StreamK.py

```diff
@@ -0,0 +1,1954 @@
+        module.add(writer.longBranchScc0(skPartialsLabel, posNeg=1))
+        # module.add(SCBranchSCC0(labelName=skPartialsLabel.getLabelName(), comment="Branch if not start tile, store partials"))
+
+        if kernel["DebugStreamK"] & 1 == 0:
+            # if we started and finished the tile, regular store code
+            # branch to regular store code, skip fixup step
+            module.add(SCmpEQU32(src0=sgpr("StreamKLocalEnd"), src1=sgpr("ItersPerTile"), comment="does wg finish tile?"))
+            module.add(SCBranchSCC1(labelName=skStoreLabel.getLabelName(), comment="Branch if started and finished tile, go to regular store code"))
+
+            # if we started the tile but did not finish it, fix up step
+            # run fixup code before regular store code
+            sCtaIdx = writer.sgprPool.checkOut(1, "CtaIdx", preventOverflow=0) # self.defineSgpr("CtaIdx", 1)
+            module.add(SAddU32(dst=sgpr(sCtaIdx), src0=sgpr("StreamKIdx"), src1=1, comment="input partial tile index"))
+
+            sFixupEnd = writer.sgprPool.checkOut(1, "FixupEnd", preventOverflow=0) # self.defineSgpr("CtaEnd", 1)
+            module.add(sMagicDivAlg2(tmpSgpr, sgpr("StreamKIterEnd"), sgpr("MagicNumberItersPerTile"), sgpr("MagicShiftItersPerTile")))
+            module.add(SMulI32(dst=sgpr(tmpSgpr), src0=sgpr(tmpSgpr), src1=sgpr("ItersPerTile"), comment="start iteration of partial tile"))
+            module.add(SSubU32(dst=sgpr(sFixupEnd), src0=sgpr("StreamKIterEnd"), src1=sgpr(tmpSgpr), comment="calc iterations completed by this WG"))
+
+            module.add(skFixupLabel)
+
+            # Check flag
+            module.add(SLShiftLeftB32(dst=sgpr(tmpSgpr), src=sgpr(sCtaIdx), shiftHex=log2(4), comment="flag offset based on CTA index"))
+            module.add(SLoadB32(dst=sgpr(tmpSgpr+2), base=sgpr("AddressFlags", 2), soffset=sgpr(tmpSgpr), smem=SMEMModifiers(glc=1), comment="get flag"))
+
+            module.add(SWaitCnt(lgkmcnt=0, comment="wait for flag load"))
+            if kernel["DebugStreamK"] & 2 == 0:
+                module.add(SCmpEQU32(src0=sgpr(tmpSgpr+2), src1=1, comment="check if ready"))
+                module.add(SCBranchSCC0(labelName=skFixupLabel.getLabelName(), comment="if flag not set, wait and check again"))
+
+            # TODO Barrier here to sync all threads in workgroup, but maybe better to have separate flag for each wavefront (to be tested)
+            module.add(SBarrier(comment="wait for all workgroups before resetting flag"))
+            skipFlagReset = Label(label=writer.labels.getNameInc("SK_SkipFlagReset"), comment="")
+            module.add(VReadfirstlaneB32(dst=sgpr(tmpSgpr+2), src=vgpr("Serial"), comment="Wave 0 updates flags"))
+            module.add(SCmpEQU32(src0=sgpr(tmpSgpr+2), src1=0, comment="Check for wave 0"))
+            module.add(SCBranchSCC0(labelName=skipFlagReset.getLabelName(), comment="Skip flag reset"))
```

Comment:
Quick question.
Why we need this code only for tensilelite?
I did not see it in Tensile.

Reply:
In Tensile the synchronization flags are stored in the workspace, which is supplied by the user and could contain any values, so there is a pre-kernel that initializes the flags to 0.
In TensileLite the flags are stored in the Synchronizer buffer which is kept in the handle and initialized to 0. In this case we also have to reset the flags to 0 after it they have been read so that they're ready for the next kernel to run. The first version on this PR had all 4 wavefronts in a workgroup doing the s_store to the flag, which was creating a race condition between the WG doing the partial tile setting it to 1 and the WG doing the fixup resetting it to 0. The kernel would sometimes end with some flags set to 1 instead of 0 causing kernels that run afterwards to fail.
This update changes it so only wavefront 0 updates the flag value (and wavefronts in a WG are still synchronized with a barrier to ensure all partial data is fully written before setting the flag)

Reply:
I just added one more test file based on the test that most easily reproduced the race condition